[WIP] wallet: use BlockFilterIndex in ScanForWalletTransactions

pstratem opened the pull request:
goal is to improve rescan performance significantly, lots more work to do

Commit:
wallet: add ScriptPubKeyMan::GetScriptPubKeys virtual function

LegacyScriptPubKey::GetScriptPubKeys is a stub

Commit:
wallet: partially implement LegacyScriptPubKeyMan::GetScriptPubKeys

implements p2pkh

MarcoFalke commented:
Have you seen #15845 ? It might be easier to take that code and then fix up the two (or three) bugs.

MarcoFalke commented:
Concept ACK either way

pstratem commented:
@MarcoFalke I hadn't, it seems you took basically the same approach that I intend to.  I'll certainly review your PR and read the comments.

Commit:
chain: add filterMatchesAny to chain interface

taken from #15845

Commit:
wallet: Use blockfilterindex "basic" to accelerate rescan time

we avoid issues with the keypool expanding during a rescan with the
script_pub_key_filter_set flag

pstratem commented:
this actually works for p2pkh transactions as it is now, benchmarks incoming

pstratem commented:
some rescanblockchain benchmarks, from my laptop:

42 minutes on master
20 minutes using the BASIC filter
4  minutes using a filter with fixed siphash parameters and avoiding recalculating the siphashes for every block

jonatack commented:
Concept ACK

gmaxwell commented:
4 minutes is awesome. 20 minutes is kinda sad compared to 42 (only 2x the speed in exchange for a bunch of disk space?).  42 minutes is *a lot* faster than what I was benchmarking before, so hurrah for whatever prior change made rescanning faster.

ryanofsky reviewed (commented):
Concept & approach ACK. Seems to pick up nicely where #15845 left off, updated to work for descriptor and legacy wallets

jb55 commented:
Concept ACK *as I grumpily wait while my wallets are rescanning*

fanquake commented on src/interfaces/chain.cpp:
```diff
@@ -186,6 +187,20 @@ class ChainImpl : public Chain
         }
         return nullopt;
     }
+    Optional<bool> filterMatchesAny(const uint256& hash, const GCSFilter::ElementSet& filter_set) override
```
Please [use `std::optional` in new code](https://github.com/bitcoin/bitcoin/blob/master/src/optional.h#L12).

MarcoFalke commented:
Marked up for grabs, along with #15845

MarcoFalke commented:
Picked up in #25957
